ref-engine-discovery.md

```diff
@@ -0,0 +1,110 @@
+# OCI Ref-engine Discovery
+
+This is version 0.1 of this specification.
+
+To faciliate communication between image publishers and consumers, this specification defines a [ref-engine discovery](#ref-engine-discovery) protocol which publishers can used to direct consumers towards [reference engines](#ref-engine).
+Having retrieved a set of reference engines (via this and other protocols), consumers can recover a set of [Merkle roots](#merkle-root) potentially associated with a given image name.
+Merkle roots may suggest [CAS engines](#cas-engine), e.g. via a `casEngines` entry in their JSON, but that is out of scope for ref-engine discovery.
+
+## Glossary
+
+### Ref-engine discovery
+
+A service that suggests possible [ref engines](#ref-engine).
+Thie specification defienes a ref-engine discovery protocol.
+
+### CAS engine
+
+A service that provides access to [content-addressable storage][cas].
+
+### Merkle root
+
+The root node in a [Merkle tree][Merkle-tree].
+In the OCI ecosystem, Merkle links are made via [descriptors][descriptor].
+The Merkle root may be a descriptor ([media type][media-type] [`application/vnd.oci.descriptor.v1+json`][descriptor]), or it may have a different media type.
+
+### Ref engine
+
+A service that maps an image name to a set of potential [Merkle roots](#merkle-root).
+
+## Well-known URI
+
+For image names which are supported for [`host-based-image-name`](host-based-image-names.md), a ref-engine discovery object MAY be retrieved from the following [well-known URI][well-known]:
+
+    https://{host}/.well-known/oci-ref-engines
+
```

Comment:
I think you means `/.well-known/oci-ref-engines` is canonical path. Should add the language: 

Where `/.well-known/oci-ref-engines` is canonical endpoint. ref-engine will handle discovery request under only this endpoint. (my language is rough)

Reply:
Another question, how to sepcify the discovery target by consumer? e.g. consumer want to specific image by name, os, arch...

We might should define some variables, which looks like:
```
https://a.b.example.com/.well-known/oci-ref-engines?name=ubuntu&os=linux&arch=amd64
```

Reply:
> `https://a.b.example.com/.well-known/oci-ref-engines?name=ubuntu&os=linux&arch=and64`

I don't think those belong in the ref-engine discovery URI, which is for "these services know how to resolve my names".  That's unlikely to be OS- or arch-dependent.  For example, I would be surprised if someone pushed all their Windows refs to one `oci-index-template-v1` ref engine but pushed all of their Linux refs to one `docker` ref engine.  And even if they did, they could get less efficient support by listing both ref engines without OS/arch qualifiers.

I think there might be a stronger case for including `os` and `arch` in the [`oci-index-template-v1` variables][1], so you could use separate index JSON for different arch and/or OS.  But I think the current variables give you enough granularity to keep the index JSON small, and with opencontainers/runtime-spec#405 still up in the air, I'm not sure if images will continue to have a single, well-defined OS (although I think the chances are better for images than for runtime configs).  So for the moment, I'd rather leave OS/arch choices in `oci-index-template-v1` up to the usual index JSON handler.  Other ref-engine protocols are, of course, free to use that information however they see fit.


[1]: https://github.com/wking/oci-discovery/blame/d694d9612a454815112bf5b01bdc89a07ee1bd2d/index-template.md#L8

Reply:
> I think you means `/.well-known/oci-ref-engines` is canonical path.

Yeah, that definition and the adjacent bit about walking the DNS ancestors were weak.  I've taken a stab at tightening them up in d694d96 → f56204e, which also renames `oci-ref-engines` to `oci-host-ref-engines` to emphasize the host-association now covered [here][1].

[1]: https://github.com/xiekeyang/oci-discovery/blob/f56204e85f9b79b1c99f3a81671fd9f90ffd9f43/ref-engine-discovery.md#images-associated-with-a-hosts-oci-host-ref-engines